### PR TITLE
Uniform & attribute caching

### PR DESCRIPTION
**Pull request by tschw**
Fixes the issue explained in #6557, also for vertex attributes, and does so properly, that is without hackily parsing the GLSL but querying the GL for caching when the renderer needs the information. 

So `WebGLProgram.prototype.uniforms` `.attributes` and `.attributesKeys` (is the latter still needed?) are functions now. API docs update included.

Plus a fix to the editor - my test case. The first commit will be rather empty, as its essence went into dev already and this branch was rebased.

### CODE REVIEW

File: editor/js/Sidebar.Material.js [hunk 1]
```diff
@@ -642,7 +642,7 @@ Sidebar.Material = function ( editor ) {
 			'emissive': materialEmissiveRow,
 			'specular': materialSpecularRow,
 			'shininess': materialShininessRow,
-			'program': materialProgramRow,
+			'vertexShader': materialProgramRow,
```
Comment (mrdoob) [inline editor/js/Sidebar.Material.js]: Is this change intended?

File: src/renderers/WebGLRenderer.js [hunk 2]
```diff
@@ -779,14 +779,15 @@ THREE.WebGLRenderer = function ( parameters ) {
 		if ( object.hasUvs && ! object.__webglUvBuffer ) object.__webglUvBuffer = _gl.createBuffer();
 		if ( object.hasColors && ! object.__webglColorBuffer ) object.__webglColorBuffer = _gl.createBuffer();
 
+		var attributes = program.attributes();
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: I think I would name it `.getAttributes()` instead.

**Comment by mrdoob** [comment]
If they're methods now I think we should name them `.getUniforms()` and `.getAttributes()`.

**Comment by mrdoob** [comment]
Looks good otherwise!

File: editor/js/Sidebar.Material.js (hunk 1 above)
Comment (tschw) [inline editor/js/Sidebar.Material.js]: Yes, certainly. The program property is not defined right away and therefore the material editor won't show the program buttons until reloaded. The vertexShader property on the other hand is defined immediately when the visibility is determined.

**Comment by tschw** [comment]
We could also add properties with getters calling the new functions and logging warnings for a smoother transition of code written against the old interface outside of the renderer. What do you think?

**Comment by mrdoob** [comment]
True! Yes, we should add a getter to show a warning. Also, I don't think AttributesKeys are no longer needed.

**Comment by tschw** [comment]
Whoops! I misread your double negation. However, a `grep -i attributeskeys` on all source files only matches the deprecated API in `WebGLProgram`. The renderer code that used to need it has recently been changed to just use a `for .. in` loop. I noticed it when pulling up.

File: editor/js/Sidebar.Material.js (hunk 1 above)
Comment (tschw) [inline editor/js/Sidebar.Material.js]: Also, `.program` is defined for any material, not just `ShaderMaterial`. I guess that effect was also unintended as impossible to work correctly, not saying it wouldn't be great if we could plug custom shader code into any material one day :).

**Comment by mrdoob** [comment]
Ah, that was a confusing double negation. I meant that we no longer need it :+1:

File: editor/js/Sidebar.Material.js (hunk 1 above)
Comment (mrdoob) [inline editor/js/Sidebar.Material.js]: We will, we will! :D

**Comment by mrdoob** [comment]
Looks good now. Many many thanks!

**Comment by tschw** [comment]
No problem - thank you!
